### PR TITLE
jax.experimental.tree_math

### LINKED ISSUE #1012: pytree transformation
I'd like to write code like `np.sum((x - y) ** 2)` that automatically works on arbitrarily nested Python objects.

I think this could be done cleanly with a "pytree transformation" that acts on functions, somewhat similar to the existing `vmap`, e.g.,
```
@jax.autotree
def norm(x, y):
    return np.sum((x - y) ** 2)
```

Under the hood, we would need to define what "autotree" versions of each primitive function look like on a list of heterogeneously shaped arrays, e.g.,
- Vectorized functions like `+` operate over each array independently.
- In mixed tree/array operations, the array should get "broadcast" by repeating it when the structure of the tree, e.g., `{'a': x, 'b': y'} + z` -> `{'a': x + z, 'b': y + z}`.
- Reductions like `sum` sum over all the arrays to produce a scalar, unless an axis is specified.

The upside is that this could make it trivial to write functional transformations like `jax.experimental.optimizers` in a fully general way. I can think of lots of nice applications for this, e.g., for methods that solve equations.

### PR DESCRIPTION
**Pull request by shoyer**
Fixes #1012

The idea behind this PR is to make it easy to implement numerical algorithms that work on pytrees, by transforming code written to handle vectors.

For example, we could implement an ODE solver that supports pytrees with something like:
```python
import functools
from jax import lax
from jax.experimental import tree_math

@functools.partial(tree_math.wrap, vector_argnums=(1,))
def odeint_midpoint(f, y0, dt, steps):
    f = tree_math.unwrap(f, vector_argnums=(0,))
    def step_fun(i, y):
        t = i * dt
        return y + dt * f(y + dt/2 * f(y, t), t + dt/2)
    return lax.fori_loop(0, num_steps, step_fun, y0)
```
Aside from `wrap` and `unwrap`, this is exactly how you would write a simple ODE solver in JAX without support for PyTrees. We currently [do something very similar](https://github.com/google/jax-cfd/blob/a92862fb757d122e7c5eee23a3b783997a2aeafe/jax_cfd/spectral/time_stepping.py#L92) for implementing ODE solvers in JAX-CFD. The tests for this PR include a rewriten version of `jax.scipy.sparse.linalg.cg`, which is also much more readable than what we currently have in JAX.

The idea here is a simpler alternative to #3263, based on a custom pytree (`tree_math.Vector`) rather than introducing a new final-style transformation:
- The upside of this approach is that writing a custom pytree is easy and entirely decoupled from JAX's other transformations. So we get support for JAX transformations (e.g., `grad`/`jit`/`vmap`) and control flow primitives (e.g., `while_loop`) for free.
- The downside is that it won't be possible to use standard `jax.numpy` functions on `Vector` objects, unless we make `jax.numpy` aware of `tree_math` (ala #8381). Instead, I've added a few specialized helper functions like `where` and `maximum`. For now, this seems like an acceptable trade-off, given that the core use-case for tree math is to
make it easy to write new algorithms from scientific computing, for which infix arithmetic is important enough that it should easily justify minor deviations from standard `jax.numpy`.

The other change from #3263 is that I've only written a simple "Vector" class for now rather than making an attempt to support arbitrary higher dimensional arrays. This also considerably simplifies the implementation. In the future, we probably will also want to add a "Matrix" class, for methods that need to keep track of multiple vectors (e.g., ODE solvers, GMRES, L-BFGS, Lanczos, etc). I don't think we'll need support for representing pytrees as higher dimensional arrays, which don't come up very often in fundamental numerical algorithms.

### CODE REVIEW

**Comment by shoyer** [comment]
I think this is ready for review!

Design aspects that I'm not entirely sure about:

- Is it a good idea to expose `tree_math.Vector` as a public API? I think this would be convenient (especially, e.g., for type checking), but this might have backwards compatibility implications if we ever switch to JAX transformation based approaches (constructors for tracer objects are not external APIs).
- Are `tree_math.wrap`/`tree_math.unwrap` the right names for the function transforms? Also, do I have them in the right order or should their meanings be flipped? Other options: `tree_math.vectorize`/`tree_math.unvectorize`, `tree_math.map`/`tree_math.unmap` (similar to `vmap`), `tree_math.vector_to_tree`/`tree_math.tree_to_vector`, ...

**Comment by NeilGirdhar** [comment]
This is awesome.  Should we favor writing algorithms for trees then and pass them to `tree_math`?

**Comment by shoyer** [comment]
> Should we favor writing algorithms for trees then and pass them to `tree_math`?

By "favor writing algorithms for trees" did you mean to say "favor writing algorithms for vectors"?

This generally would be my favored approach. Code for vectors is much more readable than code for pytrees (mostly due to support for infix arithmetic), and under `jit` performance should be identical to writing the transformations by hand.

**Comment by NeilGirdhar** [comment]
> By "favor writing algorithms for trees" did you mean to say "favor writing algorithms for vectors"?

Yes, sorry.

> Writing algorithms for vectors would definitely be my favored approach

Sounds great.  This feature is a fantastic addition in my opinion 😄

**Comment by Edenhofer** [comment]
For a lot of algorithms one would also need an implementation of `dot` (or `vdot`) for `Vector`s. There is some ambiguity here concerning the axis of the tensors of the Vector but it would be a great addition to `matmul`.

**Comment by shoyer** [comment]
> For a lot of algorithms one would also need an implementation of `dot` (or `vdot`) for `Vector`s. There is some ambiguity here concerning the axis of the tensors of the Vector but it would be a great addition to `matmul`.

The version of "matmul" that I implemented only work on vectors, so it's exactly the same as `np.dot`. So we could either include an alias `dot = matmul` or encourage using the infix notation `u @ v`.

**Comment by Edenhofer** [comment]
Sorry for the noise! I was too hasty when looking at the implementation :|
I think an alias would be nice but I don't have any strong feelings about any of the two options.

**Comment by shoyer** [comment]
I'll add a `size` property and `dot` method, those are basically aliases for existing functionality.

I'm not sure `__getitem__` makes sense. Indexing like `x[y]` on a vector `x` could only work when the indexing object `y` is a Python integer:
- There's no unambiguous way to know the underlying pytree structure for slicing like `x[:3]`.
- Indexing with a JAX tracer doesn't work, because we need to know which array is being indexed from Python to generate do efficient indexing. This means we couldn't looping over an index variable, e.g., with `lax.fori_loop`.
- This leaves indexing like `x[0]` or `x[-1]`. I can see this being handy for debugging, but I can also see it misleadingly suggesting that general purpose indexing work.

I'm not sure what you're using `__contains__` for, but generally that overload does not make sense on multi-dimensional arrays. It also needs to return a Python builtin `bool`, which is rather restrictive. JAX does not implement it on its other array types and so I doubt it makes sense here, either.

**Comment by Edenhofer** [comment]
Great, thanks!

Is `norm` also of interest or is it already too specific? It comes in handy for implementing convergence criteria in some minimization schemes which use e.g. the 1-norm or allow the user to set the order of the norm freely.

Sure, in general slicing does not work nor would a loop over all elements work efficiently. However, I don't use it for that. Most of the thinks I wrap end up being nested dictionaries of some sort and I forward `__contains__` and `__getitem__` on to the underlying object. Thus, `math_enabled_dict[some_key]` still accesses the underlying dictionary and yields whatever is at `some_key`.

My philosophy behind this decision was that whatever I wrap in a math-enabled pytree should still behave and act exactly as the un-wrapped object with the only constrain being that whatever method I implemented must also work for numpy arrays. For my implementation this means that `Vector(some_jnp_array)` behaves exactly the same as its un-wrapper counterpart. This might be overly specific to my use-case and might not fit the here-proposed `Vector` class. If so, feel free to disregard my comments on this topic.


>  JAX does not implement it on its other array types and so I doubt it makes sense here, either.

I am not sure what you mean by this; `assert 3 in jnp.arange(4)`  does work.

**Comment by shoyer** [comment]
> Is `norm` also of interest or is it already too specific? It comes in handy for implementing convergence criteria in some minimization schemes which use e.g. the 1-norm or allow the user to set the order of the norm freely.

`norm` is not a method on numpy arrays. My current approach is that the functionality of `tree_math.Vector` should be a strict subset of the functionality of normal JAX/NumPy arrays.

> Sure, in general slicing does not work nor would a loop over all elements work efficiently. However, I don't use it for that. Most of the thinks I wrap end up being nested dictionaries of some sort and I forward `__contains__` and `__getitem__` on to the underlying object. Thus, `math_enabled_dict[some_key]` still accesses the underlying dictionary and yields whatever is at `some_key`.

Ah, I see. I'd like tree_math to be entirely explicit, and _strictly_ allow for manipulating pytrees as if they were 1D NumPy arrays. If you want to do this sort of access, you can pull out the underlying Pytree with the `.tree` attribute and then make a new `Vector`.

> > JAX does not implement it on its other array types and so I doubt it makes sense here, either.
> 
> I am not sure what you mean by this; `assert 3 in jnp.arange(4)` does work.

This is Python's fallback implementation of `__contains__`, via iteration with `__iter__`: https://docs.python.org/3/reference/datamodel.html#object.__contains__

We could certainly support `__iter__` on `tree_math.Vector`. It won't be very efficient, but I guess not much worse than iterating over normal JAX arrays.

**Comment by Edenhofer** [comment]
> `norm` is not a method on numpy arrays. My current approach is that the functionality of `tree_math.Vector` should be a strict subset of the functionality of normal JAX/NumPy arrays.

Right, I got that from your implementation. Just to be clear, my proposal was to add `norm` as an additional function to `experimental.tree_math` in the same way `matmul` is exposed as function as well.

I think you also misunderstood this aspect of the proposed `size` function. Of course having it as a method of `Vector` works too but I think it would also be a great addition to `experimental.tree_math`. Similar to `matmul`, `size` and `norm` both make sense for general pytrees.

> Instead you could use infix arithmetic like (x @ x) ** 0.5.

Sure, I can of course implement any desired p-norm by hand but I find `norm(x, ord=p)` to be more convenient.

> This is Python's fallback implementation of __contains__, via iteration with __iter__: https://docs.python.org/3/reference/datamodel.html#object.__contains__

Ah, I see. Thank for the explanation!

File: jax/experimental/tree_math.py [hunk 1]
```diff
@@ -131,13 +139,17 @@ def tree_flatten(self):
   def tree_unflatten(cls, _, args):
     return cls(*args)
 
-  def __len__(self):
+  @property
+  def size(self):
     values = tree_util.tree_leaves(self.tree)
     return sum(prod(jnp.shape(value)) for value in values)
```
Comment (Edenhofer) [inline jax/experimental/tree_math.py:145]: I am asking just out of curiosity; is there a specific reason why you prefer `prod(jnp.shape(v))` over `jnp.size`?

Comment (shoyer) [reply jax/experimental/tree_math.py:145]: No, they are entirely equivalent

**Comment by shoyer** [comment]
> Similar to `matmul`, `size` and `norm` both make sense for general pytrees.

Just to be entirely clear, the functions implemented in tree_math don't work for general pytrees -- they only work for explicit `tree_math.Vector` objects. The only reason why I created `matmul` is that it gives a way to explicitly set the `precision` argument, which you can't do with infix notation. In general, I would rather encourage using methods than functions, because the methods will let you write portable code that also works on regular JAX arrays.

`norm` seems fine to add, but let's save that for a follow-on PR after this gets merged.

**Comment by Edenhofer** [comment]
Ok. Thank you for all the explanations and of course for implementing it in the first place :)

**Review by PhilipVinc — CHANGES_REQUESTED** [review]
Quantum Physicists would love `.real` and `.imag` being defined.
I think those are well defined and could be included

**Comment by PhilipVinc** [comment]
thanks!

**Comment by shoyer** [comment]
We decided to implement this as a separate library outside of JAX for now. Please take a look https://github.com/google/tree-math/ !
